File: src/java/br/com/lett/crawlernode/crawlers/ranking/keywords/brasil/BrasilLojasredeCrawler.java
package br.com.lett.crawlernode.crawlers.ranking.keywords.brasil;

import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import br.com.lett.crawlernode.core.session.Session;
import br.com.lett.crawlernode.core.task.impl.CrawlerRankingKeywords;
import br.com.lett.crawlernode.util.CommonMethods;

public class BrasilLojasredeCrawler extends CrawlerRankingKeywords{

	public BrasilLojasredeCrawler(Session session) {
		super(session);
	}

	@Override
	protected void extractProductsFromCurrentPage() {
		this.log("Página "+ this.currentPage);

		this.pageSize = 24;

		String url = "http://busca.lojasrede.com.br/busca?q=" + this.keywordEncoded + "&page="+this.currentPage;
		this.log("Link onde são feitos os crawlers: "+url);	

		this.currentDoc = fetchDocument(url);
		Elements products =  this.currentDoc.select("li.nm-product-item");

		if(products.size() >= 1) {
			if(this.totalProducts == 0) {
				setTotalProducts();
			}

			for(Element e: products) {
				String internalPid 	= crawlInternalPid(e);
				String internalId = crawlInternalId(e);
				String productUrl  = crawlProductUrl(e);

				saveDataProduct(internalId, internalPid, productUrl);

				this.log("Position: " + this.position + " - InternalId: " + internalId + " - InternalPid: " + internalPid + " - Url: " + productUrl);
				if(this.arrayProducts.size() == productsLimit) {
					break;
				}

			}
		} else {
			this.result = false;
			this.log("Keyword sem resultado!");
		}

		this.log("Finalizando Crawler de produtos da página "+this.currentPage+" - até agora "+this.arrayProducts.size()+" produtos crawleados");

	}

	@Override
	protected boolean hasNextPage() {
		return this.totalProducts > this.arrayProducts.size();
	}

	@Override
	protected void setTotalProducts() {
		Element totalElement = this.currentDoc.select(".neemu-total-products-container span").first();
		if(totalElement != null) {
			try {
				this.totalProducts = Integer.parseInt(totalElement.ownText().replaceAll("[^0-9]", ""));
			} catch(Exception e) {
				this.logError(CommonMethods.getStackTrace(e));
			}
		}

		this.log("Total da busca: " + this.totalProducts);
	}
	
	private String crawlInternalId(Element e) {
		String id = null;
		
		Element comprar = e.select(".nm-add a[data-produtovarianteid]").first();
		Element indisponivel = e.select(".spotIndisponivel .input").first();
		
		if(comprar != null) {
			id = comprar.attr("data-produtovarianteid");
		} else if(indisponivel != null) {
			String[] tokens = indisponivel.attr("id").split("-");
			id = tokens[tokens.length-1].trim();
		}

		return id;
	}
	
	private String crawlInternalPid(Element e) {
		String[] tokens = e.attr("id").split("-");
		
		return tokens[tokens.length-1].trim();
	}

	private String crawlProductUrl(Element e){
		String productUrl = null;
		Element url = e.select(".nm-product-name a").first();
		
		if(url != null) {
			productUrl = url.attr("href");
			
			if(productUrl.contains("lojasrede") && !productUrl.startsWith("http")) {
				productUrl = "http:" + productUrl;
			}
			
			if(!productUrl.startsWith("http://www.lojasrede.com.br/")) {
				productUrl = ("http://www.lojasrede.com.br/" + productUrl).replace("br//", "br/");
			}
		}
		
		return productUrl;
	}
}

File: src/java/br/com/lett/crawlernode/crawlers/ranking/keywords/brasil/BrasilLojasredeCrawler.kt
package br.com.lett.crawlernode.crawlers.ranking.keywords.brasil

import br.com.lett.crawlernode.core.session.Session
import br.com.lett.crawlernode.core.task.impl.CrawlerRankingKeywords

class BrasilLojasredeCrawler(session: Session) : CrawlerRankingKeywords(session) {

   init {
      pageSize = 24
   }

   override fun extractProductsFromCurrentPage() {
      val url = "https://www.lojasrede.com.br/${keywordWithoutAccents.replace(" ", "%20")}?PageNumber=$currentPage"
      log("Link onde são feitos os crawlers: $url")
      currentDoc = fetchDocument(url, cookies)
      val products = currentDoc.select(".vitrine.resultItemsWrapper li[layout] span[data-id]")
         .distinctBy { element -> element.attr("data-id") }
      if (products.isNotEmpty()) {
         if (totalProducts == 0) {
            setTotalProducts()
         }
         products.forEach { e ->
            val internalPid = e.attr("data-id")
            val productUrl = e.selectFirst(".product-name a")?.attr("href")
            saveDataProduct(null, internalPid, productUrl)

            log("Position: $position - InternalPid: $internalPid - Url: $productUrl")
         }
      }
      log("Pag $currentPage | ${arrayProducts.size} products")
   }

   override fun setTotalProducts() {
      val totalElement = currentDoc.selectFirst(".resultado-busca-numero .value")
      if (totalElement != null) {
         val text = totalElement.text().replace("[^0-9]".toRegex(), "").trim()
         if (text.isNotEmpty()) {
            totalProducts = text.toInt()
         }
         log("Total da busca: $totalProducts")
      }
   }
}
